enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/SlottedPipeBuilder.scala

```diff
@@ -351,7 +352,79 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
         val refOffsets = refIds.map(e => pipeline.getReferenceOffsetFor(e.name))
         ConditionalApplySlottedPipe(lhs, rhs, longOffsets, refOffsets, negated = true, pipeline)(id)
 
+      case Union(_, _) =>
+        val lhsInfo = pipelines(lhs.id)
+        val rhsInfo = pipelines(rhs.id)
+        //val lhsOverlap = overlaps(pipeline, lhsInfo) // since pipeline is constructed from lhs the overlap will always be true
+        val rhsOverlap = overlaps(pipeline, rhsInfo)
+        if(rhsOverlap == true)
+          FastUnionSlottedPipe(lhs, rhs)(id = id)
+        else
+          UnionSlottedPipe(lhs, rhs, rhsInfo, pipeline, computeUnionSlots(lhsInfo, rhsInfo, pipeline))(id = id)
+
       case _ => throw new CantCompileQueryException(s"Unsupported logical plan operator: $plan")
     }
   }
+
+  private def overlaps(source: PipelineInformation, toCheck: PipelineInformation):Boolean = {
+    source.mapSlot {
+      case (k, s) => s == toCheck.get(k).get
```

Comment:
I wonder if == is a too strict condition here? E.g. if one side has the type of a RefSlot as a CTString and the other as CTAny they could still be used as is with the fast Union? Also nullable should not matter.